File: skills/aiven-kafka-setup-avn/templates/producer_consumer_python/consumer.py
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements. See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License. You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import argparse
import csv
import logging
import os
import signal
from datetime import datetime, timezone
from typing import IO, List, Optional

from confluent_kafka import Consumer, KafkaException
from confluent_kafka.schema_registry import SchemaRegistryClient
from confluent_kafka.schema_registry.avro import AvroDeserializer
from confluent_kafka.serialization import (
    MessageField,
    SerializationContext,
    SerializationError,
)

from order import Order

TOPIC = "orders"

# Reason: 1 second balances responsiveness with network overhead on a managed
# cloud service. Shorter polls increase CPU usage; longer polls delay completion.
POLL_TIMEOUT = 1.0

DEFAULT_EXPECTED_MESSAGES = 20

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(name)s - %(message)s",
)
logger = logging.getLogger(__name__)


class OrderConsumer:
    """Consumes orders from the Kafka ``orders`` topic and writes them to a CSV.

    Use as a context manager — the underlying Kafka consumer is created on
    ``__enter__`` and closed on ``__exit__``::

        with OrderConsumer(...) as consumer:
            consumer.consume(output_path)
    """

    def __init__(
        self,
        bootstrap_servers: str,
        schema_registry_url: str,
        username: str,
        password: str,
        ca_location: str,
    ) -> None:
        self.bootstrap_servers = bootstrap_servers
        self.schema_registry_url = schema_registry_url
        self.username = username
        self.password = password
        self.ca_location = ca_location
        self.running = True
        self._consumer: Optional[Consumer] = None
        self._deserializer: Optional[AvroDeserializer] = None

    def __enter__(self) -> "OrderConsumer":
        """Create the Kafka consumer and Avro deserializer."""
        sr_client = SchemaRegistryClient(self._schema_registry_config())
        self._deserializer = AvroDeserializer(sr_client)
        self._consumer = Consumer(self._consumer_config())
        self._consumer.subscribe([TOPIC])
        return self

    def __exit__(self, *args: object) -> None:
        if self._consumer is not None:
            self._consumer.close()

    def _schema_registry_config(self) -> dict:
        # Schema Registry endpoint uses a public Let's Encrypt cert, not
        # the Aiven internal CA (cert/ca.pem). Omitting ssl.ca.location lets
        # confluent_kafka fall back to certifi's public root bundle — correct here.
        # ssl.ca.location is only needed for the broker SASL_SSL connection below.
        return {
            "url": self.schema_registry_url,
            "basic.auth.user.info": f"{self.username}:{self.password}",
        }

    def _consumer_config(self) -> dict:
        return {
            "bootstrap.servers": self.bootstrap_servers,
            "security.protocol": "SASL_SSL",
            # Disable TLS hostname verification — Aiven certificates use the
            # service hostname, not the individual broker hostname.
            # librdkafka uses "none" (not "") to disable; empty string is invalid.
            "ssl.endpoint.identification.algorithm": "none",
            "sasl.mechanism": "SCRAM-SHA-256",
            "sasl.username": self.username,
            "sasl.password": self.password,
            "ssl.ca.location": self.ca_location,
            "group.id": "orders-consumer-group",
            # Static group membership: broker skips rebalance on restart if the
            # instance ID is unchanged, reassigning the same partitions immediately.
            "group.instance.id": "consumer-1",
            "auto.offset.reset": "earliest",
            "enable.auto.commit": False,
            **self._performance_config(),
        }

    def _performance_config(self) -> dict:
        return {
            # Min bytes broker accumulates before responding to a fetch (default 1).
            # Higher = fewer round-trips.
            # https://kafka.apache.org/42/configuration/consumer-configs/#consumerconfigs_fetch.min.bytes
            "fetch.min.bytes": 1,
            # Max time broker waits if fetch.min.bytes not reached (default 500 ms).
            # https://kafka.apache.org/42/configuration/consumer-configs/#consumerconfigs_fetch.max.wait.ms
            "fetch.wait.max.ms": 500,
            # Max time between poll() calls before the broker triggers a rebalance.
            # Tune alongside session.timeout.ms when processing is slow.
            # https://kafka.apache.org/42/configuration/consumer-configs/#consumerconfigs_max.poll.interval.ms
            "max.poll.interval.ms": 300000,
            # Note: large record batches could lead to session timeouts.
            # If session.timeout.ms is tuned, heartbeat.interval.ms must be tuned
            # as well (keep it ≤ 1/3 of session.timeout.ms).
            "session.timeout.ms": 45000,
            "heartbeat.interval.ms": 15000,
        }

    def stop(self) -> None:
        """Signal the consume loop to exit after the current poll completes."""
        logger.info("Shutdown signal received — stopping consumer.")
        self.running = False

    def consume(
        self,
        output_path: str,
        expected_messages: int = DEFAULT_EXPECTED_MESSAGES,
    ) -> None:
        """Fetch messages in batches until ``expected_messages`` are received and write to CSV."""
        try:
            with open(output_path, "w", newline="") as f:
                writer = csv.writer(f)
                writer.writerow(["id", "user_id", "product", "price", "completed_at"])

                received = 0

                while self.running:
                    msgs = self._consumer.consume(num_messages=10, timeout=POLL_TIMEOUT)

                    if not msgs:
                        continue

                    received += self._write_records(msgs, writer, f)
                    # Reason: commit after the batch is flushed to disk so offsets
                    # are never ahead of persisted data (at-least-once guarantee).
                    self._consumer.commit(asynchronous=False)

                    if received >= expected_messages:
                        break

                f.flush()

            logger.info(
                "Shutting down. Consumed %d messages to %s.", received, output_path
            )

        except SerializationError as e:
            logger.error("Deserialization failed — check schema compatibility: %s", e)
            raise
        except KafkaException as e:
            if e.args[0].retriable():
                logger.warning("Retriable Kafka error in consumer: %s", e)
            else:
                logger.error("Fatal Kafka error in consumer: %s", e)
            raise
        except Exception as e:
            logger.error("Unexpected error in consumer: %s", e)
            raise

    def _write_records(
        self,
        msgs: List[object],
        writer: csv.writer,
        f: IO[str],
    ) -> int:
        """Deserialize a batch of messages, write each as a CSV row, and flush the file.
        """
        count = 0
        for msg in msgs:
            if msg.error():
                logger.error("Consumer error: %s", msg.error())
                continue
            order = Order.from_avro(
                self._deserializer(
                    msg.value(),
                    SerializationContext(TOPIC, MessageField.VALUE),
                )
            )
            completed_at = datetime.now(timezone.utc).strftime("%d-%m-%YZ%H:%M:%S")
            row = [order.id, order.user_id, order.product, order.price, completed_at]
            writer.writerow(row)
            count += 1
            logger.info("Consumed: %s", row)
        f.flush()
        return count


def main() -> None:
    parser = argparse.ArgumentParser(description="Consume orders from Kafka and write to CSV.")
    parser.add_argument("output", nargs="?", default="orders_completed.csv", help="Output CSV path")
    parser.add_argument(
        "--max-messages",
        type=int,
        default=DEFAULT_EXPECTED_MESSAGES,
        help="Stop after consuming this many messages (default: %(default)s)",
    )
    args = parser.parse_args()

    with OrderConsumer(
        bootstrap_servers=f"{os.environ['KAFKA_HOST']}:{os.environ['KAFKA_PORT']}",
        schema_registry_url=os.environ["SCHEMA_REGISTRY_URL"],
        username="consumer-user",
        password=os.environ["CONSUMER_PASSWORD"],
        ca_location="cert/ca.pem",
    ) as order_consumer:
        signal.signal(signal.SIGTERM, lambda _s, _f: order_consumer.stop())
        signal.signal(signal.SIGINT, lambda _s, _f: order_consumer.stop())
        order_consumer.consume(args.output, expected_messages=args.max_messages)


if __name__ == "__main__":
    main()

File: skills/aiven-kafka-setup-avn/templates/producer_consumer_java/Consumer.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.aiven.demo;

// Note: Confluent Serdes is used here, but alternatives like Apicurio Serdes exist.
// Apicurio provides compatibility with Confluent's API but has some limitations
// (e.g., lacks support for schemas with references when using the compatible API).
// See: https://croz.net/serde-libraries-comparison/
import io.confluent.kafka.serializers.KafkaAvroDeserializer;
import io.confluent.kafka.serializers.KafkaAvroDeserializerConfig;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.time.Duration;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Collections;
import java.util.Properties;
import java.util.concurrent.atomic.AtomicBoolean;
import org.apache.avro.generic.GenericRecord;
import org.apache.kafka.clients.consumer.CommitFailedException;
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.consumer.ConsumerRecords;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.common.errors.SerializationException;
import org.apache.kafka.common.errors.WakeupException;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/** Consumes orders from the Kafka {@code orders} topic and writes them to a CSV file. */
public final class Consumer {

  private static final Logger logger = LoggerFactory.getLogger(Consumer.class);
  private static final String TOPIC = "orders";
  private static final DateTimeFormatter FMT =
      DateTimeFormatter.ofPattern("dd-MM-yyyy'Z'HH:mm:ss");

  // Reason: 1 second balances responsiveness with network overhead on a managed cloud service.
  // Shorter polls increase CPU usage without benefit; longer polls delay completion detection.
  private static final Duration POLL_TIMEOUT = Duration.ofSeconds(1);
  private static final int EXPECTED_RECORDS = 20;

  private static final AtomicBoolean running = new AtomicBoolean(true);

  /**
   * Consumes orders and writes them to the given output file
   * (default: {@code orders_completed.csv}).
   */
  public static void main(String[] args) throws IOException {
    String bootstrapServers = System.getenv("KAFKA_HOST") + ":" + System.getenv("KAFKA_PORT");
    String schemaRegistryUrl = System.getenv("SCHEMA_REGISTRY_URL");
    String username = "consumer-user";
    String password = System.getenv("CONSUMER_PASSWORD");
    String truststorePath = System.getenv("TRUSTSTORE_PATH");
    String truststorePassword = System.getenv("TRUSTSTORE_PASSWORD");
    String outputPath = args.length > 0 ? args[0] : "orders_completed.csv";

    // Add a graceful shutdown hook to stop the consumer when the JVM exits.
    Runtime.getRuntime().addShutdownHook(new Thread(() -> running.set(false)));

    String instanceId = "consumer-1";
    Properties props =
        buildConsumerProperties(
            bootstrapServers, schemaRegistryUrl, username, password,
            truststorePath, truststorePassword, instanceId);
    applyPerformanceTuning(props);

    // consumer.close() is called automatically by try-with-resources.
    try (KafkaConsumer<String, GenericRecord> consumer = new KafkaConsumer<>(props);
        PrintWriter writer = new PrintWriter(new FileWriter(outputPath))) {

      consumer.subscribe(Collections.singletonList(TOPIC));
      writer.println("id,user_id,product,price,completed_at");

      int received = 0;

      while (running.get()) {
        try {
          ConsumerRecords<String, GenericRecord> records = consumer.poll(POLL_TIMEOUT);
          if (!records.isEmpty()) {
            received += writeRecords(records, writer);
          }
        } catch (WakeupException e) {
          // Triggered by consumer.wakeup() from another thread — clean exit.
          logger.info("Consumer wakeup — shutting down.");
          break;
        } catch (CommitFailedException e) {
          // Processing exceeded max.poll.interval.ms; broker triggered a rebalance.
          // Tune max.poll.interval.ms or reduce per-poll work to fix.
          logger.warn("Offset commit failed — processing too slow, retrying", e);
          continue;
        }
        if (received >= EXPECTED_RECORDS) {
          break;
        }
      }

      logger.info("Shutting down. Consumed {} messages to {}.", received, outputPath);
    } catch (SerializationException e) {
      // Fatal: schema mismatch between producer and consumer. Verify schema in the registry.
      logger.error("Deserialization failed — check schema compatibility", e);
      throw e;
    } catch (Exception e) {
      logger.error("Unexpected error in consumer", e);
      throw e;
    }
  }

  private static Properties buildConsumerProperties(
      String bootstrapServers,
      String schemaRegistryUrl,
      String username,
      String password,
      String truststorePath,
      String truststorePassword,
      String instanceId) {
    Properties props = new Properties();
    props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
    props.put(ConsumerConfig.GROUP_ID_CONFIG, "orders-consumer-group");
    props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
    // Static group membership — if the process restarts with the same instance ID,
    // the broker skips rebalance and reassigns the same partitions immediately.
    props.put(ConsumerConfig.GROUP_INSTANCE_ID_CONFIG, instanceId);
    props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
    props.put(
        ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, KafkaAvroDeserializer.class.getName());
    props.put(KafkaAvroDeserializerConfig.SCHEMA_REGISTRY_URL_CONFIG, schemaRegistryUrl);
    props.put("basic.auth.credentials.source", "USER_INFO");
    props.put("basic.auth.user.info", username + ":" + password);
    props.put("security.protocol", "SASL_SSL");
    props.put("ssl.endpoint.identification.algorithm", "");
    props.put("sasl.mechanism", "SCRAM-SHA-256");
    props.put(
        "sasl.jaas.config",
        "org.apache.kafka.common.security.scram.ScramLoginModule required "
            + "username=\"" + username + "\" password=\"" + password + "\";");
    // Note: keystore config is not required for SASL_SSL
    props.put("ssl.truststore.type", "PKCS12");
    props.put("ssl.truststore.location", truststorePath);
    props.put("ssl.truststore.password", truststorePassword);
    return props;
  }

  private static void applyPerformanceTuning(Properties props) {
    // Min bytes broker accumulates before responding to a fetch (default 1). Higher = fewer
    // round-trips.
    // https://kafka.apache.org/42/configuration/consumer-configs/#consumerconfigs_fetch.min.bytes
    props.put(ConsumerConfig.FETCH_MIN_BYTES_CONFIG, 1);
    // Max time broker waits if fetch.min.bytes not reached (default 500ms).
    // https://kafka.apache.org/42/configuration/consumer-configs/#consumerconfigs_fetch.max.wait.ms
    props.put(ConsumerConfig.FETCH_MAX_WAIT_MS_CONFIG, 500);
    // Max time between poll() calls before the broker considers the consumer dead and rebalances.
    // Tune this alongside session.timeout.ms if processing is slow (to avoid CommitFailedException).
    // https://kafka.apache.org/42/configuration/consumer-configs/#consumerconfigs_max.poll.interval.ms
    props.put(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG, 300000);

    // Note: A large amount of records to process could lead to session timeouts.
    // If we tune session timeouts, we should tune heartbeat intervals as well.
    props.put(ConsumerConfig.SESSION_TIMEOUT_MS_CONFIG, 45000);
    props.put(ConsumerConfig.HEARTBEAT_INTERVAL_MS_CONFIG, 15000);
  }

  private static int writeRecords(
      ConsumerRecords<String, GenericRecord> records, PrintWriter writer) {
    int count = 0;
    for (var record : records) {
      GenericRecord order = record.value();
      String completedAt = LocalDateTime.now().format(FMT);
      String line =
          String.format(
              "%s,%s,%s,%s,%s",
              order.get("id"), order.get("user_id"),
              order.get("product"), order.get("price"), completedAt);
      writer.println(line);
      logger.info("Consumed: {}", line);
      count++;
      if (count % 100 == 0) {
        writer.flush();
      }
    }
    writer.flush();
    return count;
  }
}
